once/_iterator_wrappers.py

```diff
@@ -24,142 +26,232 @@ def __init__(self) -> None:
+    # practice, removing the time.sleep(0) call would result in large drops in performance when
+    # running the multithreaded unit tests.
+    @property
+    @abc.abstractmethod
+    def generating(self) -> bool:
+        raise NotImplementedError()
+
+    @generating.setter
+    @abc.abstractmethod
+    def generating(self, val: bool):
+        raise NotImplementedError()
+
+    def compute_next_action(
+        self, result: IteratorResults, i: int
+    ) -> typing.Tuple[_IteratorAction, typing.Any]:
+        """Must be called with lock."""
+        if i == len(result.items):
+            if result.finished:
+                if result.exception:
+                    raise result.exception
+                return _IteratorAction.RETURNING, None
+            if self.generating:
+                return _IteratorAction.WAITING, None
+            else:
+                # If all of these functions are called with locks, we will never have more than one
+                # caller have GENERATING at any time.
+                self.generating = True
+                return _IteratorAction.GENERATING, None
+        else:
+            return _IteratorAction.YIELDING, result.items[i]
+
+    def record_successful_completion(self, result: IteratorResults):
+        """Must be called with lock."""
+        result.finished = True
+        self.generating = False
+        self.generator = None  # Allow this to be GCed.
+        self.callable = None  # Allow this to be GCed.
+
+    def record_item(self, result: IteratorResults, item: typing.Any):
+        """Must be called with lock."""
+        self.generating = False
+        result.items.append(item)
+
+    def record_exception(self, result: IteratorResults, exception: Exception):
+        """Must be called with lock."""
+        result.finished = True
+        # We need to keep track of the exception so that we can raise it in the same
+        # position every time the iterator is called.
+        result.exception = exception
+        self.generating = False
+        assert self.callable is not None
+        self.generator = self.callable()  # Reset the iterator for the next call.
+        if self.reset_on_exception:
+            self.result = IteratorResults()
+        else:
+            self.generator = None  # allow this to be GCed
+
+
+class AsyncGeneratorWrapper(_GeneratorWrapperBase):
     """Wrapper around an async generator which only runs once.
 
     Subsequent calls will return results from the first call, which is
     evaluated lazily.
     """
 
-    def __init__(self, func, *args, **kwargs) -> None:
-        self.generator: collections.abc.AsyncGenerator | None = func(*args, **kwargs)
-        self.result = IteratorResults()
-        self.generating = False
+    generator: collections.abc.AsyncGenerator
+
+    def __init__(self, *args, **kwargs) -> None:
+        self.active_generation_completed = asyncio.Event()
+        super().__init__(*args, **kwargs)
```

Comment:
Is there any reason for interweaving super().__init__ between the two local variables? If there is, a comment might be nice, otherwise I would just call super first.

Reply:
added a comment.